Add expect_true for irrefutable guards

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106788
* __->__ #106720

Here's what it does from the comments:

```
Assume that a boolean is true for the purposes of subsequent symbolic
reasoning.  This will keep track of corresponding runtime checks to verify
that the result is upheld: either as a regular guard, or as a special set
of asserts which are triggered when an unbacked SymInt is allocated.

DO NOT use this function for these cases:

 - This is inappropriate for "branching" conditions (where both
   true and false result in valid programs).  We will always assume
   the condition evaluates true, and so it will never be possible
   to trace the false condition when you use it.  For true branching
   on unbacked SymInts, you must use torch.cond.

 - This is inappropriate for situations where you know some other system
   invariant guarantees that this property holds, since you don't
   really need to insert a runtime check in that case.  Use something
   like constrain_range in that case.

This API has a hitch.  To avoid having to reimplement error reporting
capabilities, this function CAN return False.  The invariant is that
the surrounding code must raise an error when this function returns
False.  This is quite low level, so we recommend using other functions
like check() which enforce this in a more intuitive way.

By the way, this name is a nod to the __builtin_expect likely macro,
which is used similarly (but unlike __builtin_expect, you MUST fail
in the unlikely branch.)
```

We don't do anything with this right now, except use it to discharge regular guards.  Follow up PRs to (1) use it at important error checking sites, (2) actually ensure the runtime asserts make there way into the exported IR / inductor generated code.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Add expect_true for irrefutable guards

Here's what it does from the comments:

```
Assume that a boolean is true for the purposes of subsequent symbolic
reasoning.  This will keep track of corresponding runtime checks to verify
that the result is upheld: either as a regular guard, or as a special set
of asserts which are triggered when an unbacked SymInt is allocated.

DO NOT use this function for these cases:

 - This is inappropriate for "branching" conditions (where both
   true and false result in valid programs).  We will always assume
   the condition evaluates true, and so it will never be possible
   to trace the false condition when you use it.  For true branching
   on unbacked SymInts, you must use torch.cond.

 - This is inappropriate for situations where you know some other system
   invariant guarantees that this property holds, since you don't
   really need to insert a runtime check in that case.  Use something
   like constrain_range in that case.

This API has a hitch.  To avoid having to reimplement error reporting
capabilities, this function CAN return False.  The invariant is that
the surrounding code must raise an error when this function returns
False.  This is quite low level, so we recommend using other functions
like check() which enforce this in a more intuitive way.

By the way, this name is a nod to the __builtin_expect likely macro,
which is used similarly (but unlike __builtin_expect, you MUST fail
in the unlikely branch.)
```

We don't do anything with this right now, except use it to discharge regular guards.  Follow up PRs to (1) expose this from C++, (2) use it at important error checking sites, (3) actually ensure the runtime asserts make there way into the exported IR / inductor generated code.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3275,39 +3414,62 @@ def evaluate_expr(self, orig_expr: "sympy.Expr", hint=None, fx_node=None):
-                    maybe_extra_debug,
-                    stack_info=is_debug,
-                )
+            self._log_guard(self, "eval", g, tb)
```
This is code motion only to `_log_guard`

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3231,20 +3383,7 @@ def evaluate_expr(self, orig_expr: "sympy.Expr", hint=None, fx_node=None):
-                },
-            )
-            log.warning("Ignored guard %s == %s, this could result in accuracy problems", expr, concrete_val)
+        self._check_frozen(expr, concrete_val)
```
This is code motion only to `_check_frozen`

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3175,6 +3277,56 @@ def _simplify_floor_div(self, expr):
             self.evaluate_expr(eq_expr)
         return self.simplify(expr)
 
+    # We're about to add a guard/runtime assert, check if the ShapeEnv is frozen
```
These are preexisting code, just factored out of evaluate_expr

ezyang committed:
Update on "Add expect_true for irrefutable guards"


Here's what it does from the comments:

```
Assume that a boolean is true for the purposes of subsequent symbolic
reasoning.  This will keep track of corresponding runtime checks to verify
that the result is upheld: either as a regular guard, or as a special set
of asserts which are triggered when an unbacked SymInt is allocated.

DO NOT use this function for these cases:

 - This is inappropriate for "branching" conditions (where both
   true and false result in valid programs).  We will always assume
   the condition evaluates true, and so it will never be possible
   to trace the false condition when you use it.  For true branching
   on unbacked SymInts, you must use torch.cond.

 - This is inappropriate for situations where you know some other system
   invariant guarantees that this property holds, since you don't
   really need to insert a runtime check in that case.  Use something
   like constrain_range in that case.

This API has a hitch.  To avoid having to reimplement error reporting
capabilities, this function CAN return False.  The invariant is that
the surrounding code must raise an error when this function returns
False.  This is quite low level, so we recommend using other functions
like check() which enforce this in a more intuitive way.

By the way, this name is a nod to the __builtin_expect likely macro,
which is used similarly (but unlike __builtin_expect, you MUST fail
in the unlikely branch.)
```

We don't do anything with this right now, except use it to discharge regular guards.  Follow up PRs to (1) expose this from C++, (2) use it at important error checking sites, (3) actually ensure the runtime asserts make there way into the exported IR / inductor generated code.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add expect_true for irrefutable guards"


Here's what it does from the comments:

```
Assume that a boolean is true for the purposes of subsequent symbolic
reasoning.  This will keep track of corresponding runtime checks to verify
that the result is upheld: either as a regular guard, or as a special set
of asserts which are triggered when an unbacked SymInt is allocated.

DO NOT use this function for these cases:

 - This is inappropriate for "branching" conditions (where both
   true and false result in valid programs).  We will always assume
   the condition evaluates true, and so it will never be possible
   to trace the false condition when you use it.  For true branching
   on unbacked SymInts, you must use torch.cond.

 - This is inappropriate for situations where you know some other system
   invariant guarantees that this property holds, since you don't
   really need to insert a runtime check in that case.  Use something
   like constrain_range in that case.

This API has a hitch.  To avoid having to reimplement error reporting
capabilities, this function CAN return False.  The invariant is that
the surrounding code must raise an error when this function returns
False.  This is quite low level, so we recommend using other functions
like check() which enforce this in a more intuitive way.

By the way, this name is a nod to the __builtin_expect likely macro,
which is used similarly (but unlike __builtin_expect, you MUST fail
in the unlikely branch.)
```

We don't do anything with this right now, except use it to discharge regular guards.  Follow up PRs to (1) use it at important error checking sites, (2) actually ensure the runtime asserts make there way into the exported IR / inductor generated code.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ysiraichi commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3275,39 +3428,62 @@ def evaluate_expr(self, orig_expr: "sympy.Expr", hint=None, fx_node=None):
+            and fx_node is not None
+            and not self._suppress_guards_tls()
+        ):
+            self.create_fx_call_function(torch._assert, (fx_node,))
```
If we make it a **source** expression, then there must be a corresponding **target** expression somewhere. Since these runtime assertions are supposed to be always true, shouldn't it be simply a TV assertion or **target** expression?

ezyang replied:
Ah yes, good point. I think we need to "generate" these target expressions during produce guards.

ezyang committed:
Update on "Add expect_true for irrefutable guards"


Here's what it does from the comments:

```
Assume that a boolean is true for the purposes of subsequent symbolic
reasoning.  This will keep track of corresponding runtime checks to verify
that the result is upheld: either as a regular guard, or as a special set
of asserts which are triggered when an unbacked SymInt is allocated.

DO NOT use this function for these cases:

 - This is inappropriate for "branching" conditions (where both
   true and false result in valid programs).  We will always assume
   the condition evaluates true, and so it will never be possible
   to trace the false condition when you use it.  For true branching
   on unbacked SymInts, you must use torch.cond.

 - This is inappropriate for situations where you know some other system
   invariant guarantees that this property holds, since you don't
   really need to insert a runtime check in that case.  Use something
   like constrain_range in that case.

This API has a hitch.  To avoid having to reimplement error reporting
capabilities, this function CAN return False.  The invariant is that
the surrounding code must raise an error when this function returns
False.  This is quite low level, so we recommend using other functions
like check() which enforce this in a more intuitive way.

By the way, this name is a nod to the __builtin_expect likely macro,
which is used similarly (but unlike __builtin_expect, you MUST fail
in the unlikely branch.)
```

We don't do anything with this right now, except use it to discharge regular guards.  Follow up PRs to (1) use it at important error checking sites, (2) actually ensure the runtime asserts make there way into the exported IR / inductor generated code.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2794,6 +2885,13 @@ def issue_guard(guard: ShapeGuard) -> None:
+            # set
+            for ras in self.deferred_runtime_asserts.values():
+                for ra in ras:
+                    self._add_target_expr(ra.expr)
```
@ysiraichi Here's the fix. I can't easily test it in this PR as there are no uses of expect_true, but I'm able to show in the next PR that TV fails without this change and passes after on this test: https://github.com/pytorch/pytorch/pull/106788/files#diff-0daa64329e2d8648fc119f7809dc810c744508d345ea023116614ccc17e57dbe

ysiraichi commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3244,39 +3404,62 @@ def evaluate_expr(self, orig_expr: "sympy.Expr", hint=None, fx_node=None):
+            # TODO: refine ranges
+            # Unfortunately, range refinement is probably going to not
+            # work most of the time, because we don't support symbols
+            # in ranges.  For example, i0 <= s0 is un-rangeable, because
```
We might not be able to put `s0` in a range, but `i0 <= s0` still can refine `i0` value ranges. Suppose `i0: [a, b]` and `s0: [c, d]`. `i0 <= s0` refinement results will be: `i0: [a, min(b, d)]`.

ezyang replied:
Yeah, there might be some simple stuff possible. In the split example I was looking at, this wouldn't help though, because I typically don't have a range on s0.

ysiraichi reviewed (approved):
LGTM

ezyang commented:
@pytorchbot merge

voznesenskym commented on c10/core/SymBool.h:
```diff
@@ -52,6 +52,7 @@ class C10_API SymBool {
   // that value.  Note that C++ comparison operations default to returning
   // bool, so it's not so common to have to call this
   bool guard_bool(const char* file, int64_t line) const;
+  bool expect_true(const char* file, int64_t line) const;
```
guard_true feels more in line with (1) guard_bool (2) this not being a test (expect... feels firmly like testing nomenclature)

ezyang replied:
It's not a guard. They do strictly different things.

The expect nomenclature is inline with `__builtin_expect` which matches what is going on here exactly.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1434,6 +1481,20 @@ def wrapper(self, *args, **kwargs):
+    stack: str
+
+    def __repr__(self):
+        return str(self.expr)
```
I would expect repr to at least tell me the message and that this is an assert?

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -418,6 +418,43 @@ def constrain_unify(a, b):
+# False.  This is quite low level, so we recommend using other functions
+# like check() which enforce this in a more intuitive way.
+#
+# By the way, this name is a nod to the __builtin_expect likely macro,
```
Yes, it is stronger than likely.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -418,6 +418,43 @@ def constrain_unify(a, b):
+#
+# DO NOT use this function for these cases:
+#
+#  - This is inappropriate for "branching" conditions (where both
+#    true and false result in valid programs).  We will always assume
+#    the condition evaluates true, and so it will never be possible
+#    to trace the false condition when you use it.  For true branching
+#    on unbacked SymInts, you must use torch.cond.
```
If you incorrectly `expect_true`, this will cause the false branch to become completely unreachable (because we will never trace into it.)

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2885,14 +2983,39 @@ def get_shape_groups(self):
+                else:
+                    e = ra.expr
+                subst[e] = sympy.true
+                subst[sympy.Not(e)] = sympy.false
```
do you just need a table of relations and flipping? How hard is this to add?

ezyang replied:
It's not hard, but to be honest the use of subst to do these replacements is a hack in and of itself, so I didn't want to dig too far into the hole. Relevant Slack discuss about a better way to do this in sympy: https://pytorch.slack.com/archives/CFZ7V4DNZ/p1691504413431959

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3244,39 +3404,62 @@ def evaluate_expr(self, orig_expr: "sympy.Expr", hint=None, fx_node=None):
+        # NB: Don't use new_expr as expr; it could contain gunk like shape0
+        # which we don't want to guard on
+
+        # OK, we're definitely doing a runtime assert now
```
debug log? Feels useful to know when it fell through to this!

ezyang replied:
this is covered by

```
            self._log_guard("runtime_assert", expr, tb)
        else:
            self.log.debug("runtime_assert %s [guard suppressed]", expr)
```

which covers all non-exceptional branches

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3244,39 +3404,62 @@ def evaluate_expr(self, orig_expr: "sympy.Expr", hint=None, fx_node=None):
+            stack = ''.join(traceback.format_list(tb))
+            ra = RuntimeAssert(expr, msg, stack)
+            # TODO: Do this in a way that is less janky than int(s.name[1:])
+            cands = sorted([s for s in expr.free_symbols if s.name.startswith("i")], key=lambda s: int(s.name[1:]))
```
What about symints that don't start with `i`? We have those in inductor, no? Or do we expect them not to his this codepath?

ezyang replied:
You only want to match unbacked SymInts. This is sound as long as no one starts allocating symbols that also start with i, but aren't unbacked. I'm not very happy with the symbol names having meaning but there's not really other convenient ways of testing this (well, maybe I guess I could maintain a set of unbacked symbols...)

ezyang committed:
Update on "Add expect_true for irrefutable guards"


Here's what it does from the comments:

```
Assume that a boolean is true for the purposes of subsequent symbolic
reasoning.  This will keep track of corresponding runtime checks to verify
that the result is upheld: either as a regular guard, or as a special set
of asserts which are triggered when an unbacked SymInt is allocated.

DO NOT use this function for these cases:

 - This is inappropriate for "branching" conditions (where both
   true and false result in valid programs).  We will always assume
   the condition evaluates true, and so it will never be possible
   to trace the false condition when you use it.  For true branching
   on unbacked SymInts, you must use torch.cond.

 - This is inappropriate for situations where you know some other system
   invariant guarantees that this property holds, since you don't
   really need to insert a runtime check in that case.  Use something
   like constrain_range in that case.

This API has a hitch.  To avoid having to reimplement error reporting
capabilities, this function CAN return False.  The invariant is that
the surrounding code must raise an error when this function returns
False.  This is quite low level, so we recommend using other functions
like check() which enforce this in a more intuitive way.

By the way, this name is a nod to the __builtin_expect likely macro,
which is used similarly (but unlike __builtin_expect, you MUST fail
in the unlikely branch.)
```

We don't do anything with this right now, except use it to discharge regular guards.  Follow up PRs to (1) use it at important error checking sites, (2) actually ensure the runtime asserts make there way into the exported IR / inductor generated code.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

bdhirsh commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2798,6 +2894,13 @@ def issue_guard(guard: ShapeGuard) -> None:
+            # Add all deferred runtime assertions; these are not technically
+            # handled by produce_guards but we need to put them in the target
+            # set
+            for ras in self.deferred_runtime_asserts.values():
```
Basic question: in order for this PR to be sound, doesn't the resulting compiled code that dynamo eventually runs need to include code that executes the runtime asserts that we've added to the shape env? Where is that happening in this PR?

Related q: you mentioned in the PR description that there's a TODO to make sure that the asserts show up in in the generated code. Do we expect **inductor** to be the one generating the asserts from the shape env? Since e.g. if you run with `torch.compile(f, backend="aot_eager")`, wouldn't we still want to make sure that the asserts fire?

ezyang commented:
@pytorchbot merge -ic

ezyang commented:
@pytorchbot merge -i

ezyang committed:
Update on "Add expect_true for irrefutable guards"


Here's what it does from the comments:

```
Assume that a boolean is true for the purposes of subsequent symbolic
reasoning.  This will keep track of corresponding runtime checks to verify
that the result is upheld: either as a regular guard, or as a special set
of asserts which are triggered when an unbacked SymInt is allocated.

DO NOT use this function for these cases:

 - This is inappropriate for "branching" conditions (where both
   true and false result in valid programs).  We will always assume
   the condition evaluates true, and so it will never be possible
   to trace the false condition when you use it.  For true branching
   on unbacked SymInts, you must use torch.cond.

 - This is inappropriate for situations where you know some other system
   invariant guarantees that this property holds, since you don't
   really need to insert a runtime check in that case.  Use something
   like constrain_range in that case.

This API has a hitch.  To avoid having to reimplement error reporting
capabilities, this function CAN return False.  The invariant is that
the surrounding code must raise an error when this function returns
False.  This is quite low level, so we recommend using other functions
like check() which enforce this in a more intuitive way.

By the way, this name is a nod to the __builtin_expect likely macro,
which is used similarly (but unlike __builtin_expect, you MUST fail
in the unlikely branch.)
```

We don't do anything with this right now, except use it to discharge regular guards.  Follow up PRs to (1) use it at important error checking sites, (2) actually ensure the runtime asserts make there way into the exported IR / inductor generated code.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
